src/dotty/tools/dotc/core/TypeApplications.scala

```diff
@@ -500,121 +419,133 @@ class TypeApplications(val self: Type) extends AnyVal {
-   *  2. If `T` is some other type lambda `[X1,...,Xn] -> S` none of the arguments
-   *     `U1,...,Un` is a wildcard, compute `[X1:=U1, ..., Xn:=Un]S` instead.
-   *  3. If `T` is a polytype, instantiate it to `U1,...,Un`.
    */
   final def appliedTo(args: List[Type])(implicit ctx: Context): Type = /*>|>*/ track("appliedTo") /*<|<*/ {
-    def substHkArgs = new TypeMap {
-      def apply(tp: Type): Type = tp match {
-        case TypeRef(RefinedThis(rt), name) if rt.eq(self) && name.isHkArgName =>
-          args(name.hkArgIndex)
-        case _ =>
-          mapOver(tp)
-      }
-    }
-    if (args.isEmpty || ctx.erasedTypes) self
-    else self.stripTypeVar match {
-      case EtaExpansion(self1) =>
-        self1.appliedTo(args)
-      case TypeLambda(_, _, body) if !args.exists(_.isInstanceOf[TypeBounds]) =>
-        substHkArgs(body)
-      case self: PolyType =>
-        self.instantiate(args)
-      case self1 =>
-        self1.safeDealias.appliedTo(args, typeParams)
-    }
-  }
-
-  /** Encode application `T[U1, ..., Un]` without simplifications, where
-   *  @param self     = `T`
-   *  @param args     = `U1, ..., Un`
-   *  @param tparams  are assumed to be the type parameters of `T`.
-   */
-  final def appliedTo(args: List[Type], typParams: List[TypeSymbol])(implicit ctx: Context): Type = {
-    def matchParams(t: Type, tparams: List[TypeSymbol], args: List[Type])(implicit ctx: Context): Type = args match {
+    val typParams = self.typeParams
+    def matchParams(t: Type, tparams: List[TypeParamInfo], args: List[Type])(implicit ctx: Context): Type = args match {
       case arg :: args1 =>
         try {
           val tparam :: tparams1 = tparams
-          matchParams(RefinedType(t, tparam.name, arg.toBounds(tparam)), tparams1, args1)
+          matchParams(RefinedType(t, tparam.paramName, arg.toBounds(tparam)), tparams1, args1)
         } catch {
           case ex: MatchError =>
-            println(s"applied type mismatch: $self $args, typeParams = $typParams") // !!! DEBUG
+            println(s"applied type mismatch: $self with underlying ${self.underlyingIfProxy}, args = $args, typeParams = $typParams") // !!! DEBUG
             //println(s"precomplete decls = ${self.typeSymbol.unforcedDecls.toList.map(_.denot).mkString("\n  ")}")
             throw ex
         }
       case nil => t
     }
-    assert(args.nonEmpty)
-    matchParams(self, typParams, args) match {
-      case refined @ RefinedType(_, pname) if pname.isHkArgName =>
-        TypeRef(refined, tpnme.hkApply)
-      case refined =>
-        refined
+    val stripped = self.stripTypeVar
+    val dealiased = stripped.safeDealias
+    if (args.isEmpty || ctx.erasedTypes) self
+    else dealiased match {
+      case dealiased: TypeLambda =>
+        def tryReduce =
+          if (!args.exists(_.isInstanceOf[TypeBounds])) {
+            val followAlias = stripped match {
+              case stripped: TypeRef =>
+                stripped.symbol.is(BaseTypeArg)
```

Comment:
Could you add a comment explaining why reduction is safe/needed here? Also any reason to not turn that off when `Config.simplifyApplications` is false?